feat: add transactions types

developerfred opened the pull request:
ref #270
<img width="1278" alt="Captura de Tela 2024-01-04 às 20 35 42" src="https://github.com/smlxl/evm.codes/assets/57037080/20c0fe57-7243-40b4-8a1c-5ee46007e27b">

dorlevi commented:
@developerfred : it seems like the build is failing on this branch, are you addressing it?

charisra commented on components/Reference/columns.tsx:
```diff
@@ -1,60 +1,112 @@
+              <span
+                key={index}
+                className={`ml-2 py-1 px-3 leading-normal rounded-full text-2xs tracking-widest font-medium ${
+                  rollupName === 'optimism'
+                    ? 'bg-red-500 text-white' 
+                    : rollupName === 'base'
+                    ? 'bg-blue-500 text-white'
+                    : rollupName === 'arbitrumOne'
+                    ? 'bg-blue-300 text-white'
+                    : 'bg-gray-400 text-gray-800' 
```
a bit hard to grasp with all those consecutive ternaries. Better extract.

charisra commented on components/Reference/columns.tsx:
```diff
@@ -1,60 +1,112 @@
+                    ? 'bg-blue-300 text-white'
+                    : 'bg-gray-400 text-gray-800' 
+                }`}
+                style={{ margin: '4px' }}
```
Do you need margin explicitly in styles here or can it be moved to `className` and become a tailwind class?

charisra commented on context/ethereumContext.tsx:
```diff
@@ -151,9 +154,19 @@ export const EthereumProvider: React.FC<{}> = ({ children }) => {
   }, [])
 
+  const loadTransactionTypes = () => {
+    // Transforma o JSON em um array apropriado
```
English please?

charisra commented on lib/useActions.tsx:
```diff
@@ -28,6 +71,16 @@ const useActions = () => {
+      shortcut: ['t'],
+      keywords: 'Transactions Types',
+      section: 'Navigation',
+      subtitle: 'Transactions Types reference',
```
Maybe `Transaction Types reference` ?

charisra commented on pages/transactions.tsx:
```diff
@@ -0,0 +1,130 @@
+
+const { serverRuntimeConfig } = getConfig()
+
+// It seems the memory expansion computation and constants did not change since frontier, but we have to keep an eye on new fork to keep this up to date
```
Better break this comment to 2 lines

charisra commented on pages/transactions.tsx:
```diff
@@ -0,0 +1,130 @@
+  return (
+    <>
+      <html lang="en"></html>
+      <React.Fragment>
```
Unnecessary `<React.Fragment>`

charisra commented on pages/transactions.tsx:
```diff
@@ -0,0 +1,130 @@
+  await Promise.all(
+    docs.map(async (doc) => {
+      const filePath = path.join(docsPath, doc);
+      const stat = fs.statSync(filePath);
```
Please use a better variable name than `stat`. What does this represent? It's unclear for me reading it.

charisra reviewed (commented):
Few comments, also please check the failed build.

2xic commented:
@developerfred do you need any help addressing the feedback / getting the build to complete ?

developerfred commented:
> @developerfred do you need any help addressing the feedback / getting the build to complete ?

If you could help that would be great too, I can't see the Vercel logs I'm doing typecheck locally

2xic commented:
> > @developerfred do you need any help addressing the feedback / getting the build to complete ?
> 
> If you could help that would be great too, I can't see the Vercel logs I'm doing typecheck locally

Sure, attached the command for running the typechecker and how to fix them 😄 

Let us know if you still have any issues 🙌 

## Finding the build error

If you run `yarn run typecheck` you should see the following two build errors (if you don't make sure you ran `yarn install` first)

```
brage@brages-MBP developerfred-evm.codes % yarn run typecheck

> evm.codes@0.1.0 typecheck
> tsc --pretty

components/KBar/Results.tsx:8:11 - error TS2339: Property 'results' does not exist on type 'ActionGroup[]'.

8   const { results } = useMatches()
            ~~~~~~~

pages/_app.tsx:38:27 - error TS2322: Type '({ id: string; name: string; shortcut: string[]; keywords: string; section: string; subtitle: string; icon: Element; children: { id: string; name: string; shortcut: string[]; keywords: string; section: string; perform: () => void; subtitle: string; icon: Element; }[]; perform?: undefined; } | { ...; } | { ...; })[]' is not assignable to type 'Action[]'.
  Type '{ id: string; name: string; shortcut: string[]; keywords: string; section: string; subtitle: string; icon: Element; children: { id: string; name: string; shortcut: string[]; keywords: string; section: string; perform: () => void; subtitle: string; icon: Element; }[]; perform?: undefined; } | { ...; } | { ...; }' is not assignable to type 'Action'.
    Type '{ id: string; name: string; shortcut: string[]; keywords: string; section: string; subtitle: string; icon: JSX.Element; children: { id: string; name: string; shortcut: string[]; keywords: string; section: string; perform: () => void; subtitle: string; icon: JSX.Element; }[]; perform?: undefined; }' is not assignable to type 'Action'.
      Types of property 'children' are incompatible.
        Type '{ id: string; name: string; shortcut: string[]; keywords: string; section: string; perform: () => void; subtitle: string; icon: Element; }[]' is not assignable to type 'string[]'.
          Type '{ id: string; name: string; shortcut: string[]; keywords: string; section: string; perform: () => void; subtitle: string; icon: JSX.Element; }' is not assignable to type 'string'.

38             <KBarProvider actions={actions}>
                             ~~~~~~~

  node_modules/kbar/lib/types.d.ts:27:5
    27     actions: Action[];
           ~~~~~~~
    The expected type comes from property 'actions' which is declared here on type 'IntrinsicAttributes & KBarProviderProps & { children?: ReactNode; }'


Found 2 errors in 2 files.

Errors  Files
     1  components/KBar/Results.tsx:8
     1  pages/_app.tsx:38
```


## Fixing components/KBar/Results.tsx:8
So first step would be to fix the error in `Results.tsx` .

Since `useMatches()` returns an array we want to switch from 

https://github.com/smlxl/evm.codes/blob/eb75abc3fe940276e80f0ae64db9a6dbfad13bbc/components/KBar/Results.tsx#L8

to 

```typescript
const results = useMatches() 
```

## Fixing pages/_app.tsx:38

Then fix the issue in `pages/_app.tsx` where you added a new entry for the theme

(only showing the relevant code entry, but click the link for full context)
https://github.com/smlxl/evm.codes/blob/eb75abc3fe940276e80f0ae64db9a6dbfad13bbc/lib/useActions.tsx#L13-L53

The `Action` type wants just the action id in the `children` field so you should be able to switch this to 

```typescript
    {
      id: 'theme',
      name: 'Theme',
      shortcut: ['t'],
      keywords: 'change theme',
      section: 'Settings',
      subtitle: 'Change application theme',
      icon: <Icon name="palette-line" />,
      children: [
        'theme-light',
        'theme-dark',
        'theme-system'
      ],
    },
    {
      id: 'theme-light',
      name: 'Light Mode',
      shortcut: ['l'],
      keywords: 'light theme',
      section: 'Settings',
      perform: () => setTheme('light'),
      subtitle: 'Switch to Light Mode',
      icon: <Icon name="sun-line" />,
    },
    {
      id: 'theme-dark',
      name: 'Dark Mode',
      shortcut: ['d'],
      keywords: 'dark theme',
      section: 'Settings',
      perform: () => setTheme('dark'),
      subtitle: 'Switch to Dark Mode',
      icon: <Icon name="moon-line" />,
    },
    {
      id: 'theme-system',
      name: 'System Default',
      shortcut: ['s'],
      keywords: 'system theme',
      section: 'Settings',
      perform: () => setTheme('system'),
      subtitle: 'Use system theme settings',
      icon: <Icon name="settings-2-line" />,
    },
```

2xic commented:
@developerfred do you need any more help here - I'm happy to help so we can get this in a reviewable state 😄

charisra commented:
@developerfred all you need to for the to fix the build error is replace this:
```
children: [
  'theme-light',
  'theme-dark',
  'theme-system'
],
```

with this:
`children: ['theme-light', 'theme-dark', 'theme-system']`

in: ./lib/useActions.tsx

Want to give this a try and see if it builds? I can help more if you face more problems. Would really help if you got your local ESLint properly setup, if you haven’t already.

developerfred commented:
Thanks @2xic @dorlevi  and @charisra building working now 

version working https://evm-codes-fmjmeuoa2-developerfred.vercel.app/transactions

charisra commented:
> Thanks @2xic @dorlevi and @charisra build working now
> 
> version working [evm-codes-fmjmeuoa2-developerfred.vercel.app/transactions](https://evm-codes-fmjmeuoa2-developerfred.vercel.app/transactions)

It's lot letting me preview so can't tell if it's ok @developerfred . I still see errors building in the latest builds. Can you do a check that your latest commits build successfully? 

Also, if you're using VSCode, please ensure you have [Prettier](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) installed and enabled. It should auto-fix several code formatting errors.

developerfred commented:
@charisra My commits were compiled, I also used prettier, what are the next steps?

charisra commented:
> @charisra My commits were compiled, I also used prettier, what are the next steps?

Great. I'll review once more, also check the Preview, approve and then you can merge.

charisra commented on components/Reference/Header.tsx:
```diff
@@ -6,14 +6,23 @@ import { H2, Label } from 'components/ui'
+const ReferenceHeader = ({ isPrecompiled, isTransactionType }: Props) => {
   const { selectedFork } = useContext(EthereumContext)
 
+  let title = 'Instructions'
+  if (isPrecompiled) {
+    title = 'Precompiled Contracts'
+  }
+  if (isTransactionType) {
+    title = 'Transaction Types'
+  }
```
Optional: Can become a one-liner: `let title = isPrecompiled ? 'Precompiled Contracts' : isTransactionType ? 'Transaction Types' : 'Instructions';`

charisra commented on docs/opcodes/16.mdx:
```diff
@@ -3,7 +3,7 @@ fork: Frontier
 ---
 
-*Index 1 is top of the stack. See [PUSH](/#60).*
+_Index 1 is top of the stack. See [PUSH](/#60)._
```
This is repeated all over the place. Potentially can be extract?

developerfred replied:
Yes, that’s possible

charisra commented on docs/opcodes/30.mdx:
```diff
@@ -3,22 +3,23 @@ fork: Frontier
-|--:|------:|-------:|
-| `1` | | `0x9bbfed6889322e016e0a02ee459d306fc19545d8` |
+|  \* | Input |                                       Output |
+| --: | ----: | -------------------------------------------: |
```
There are a lot of changes in these `.mdx` files, those seem to be only code style changes and don't have any real visual impact, right?

developerfred replied:
These changes and because of the prettier in the whole code, but if you prefer I can isolate that.

charisra replied:
No if prettier fixed those, that's fine, let's keep them. The must have not been done properly in the first place.

charisra reviewed (commented):
This is going on the right track, close to be merge-able. 
@developerfred please add in the PR description a short explanation about what has been done in this PR and how we (reviewers) can test & confirm that the PR achieves its purpose. 

Left a few comments, too, should be easy to address.

charisra commented:
So that's what I'm expected to see?
<img width="1728" alt="Screenshot 2024-01-31 at 17 58 01" src="https://github.com/smlxl/evm.codes/assets/8069777/c10b8f50-9e0e-4858-9f82-850d0c6d18fa">

Just the types, there's not any more info on those, it mentions `There is no reference doc for this yet. Why not contribute )` That's the expected behavior?

developerfred commented:
> `There is no reference doc for this yet. Why not contribute` That's the expected behavior?

I believe this is a call action to attract contributors

charisra commented:
@developerfred I approved the PR. Is there anything pending before you merge?
